### DISCUSSION TITLE
Pause & Resume

### DISCUSSION THREAD

**Discussion by YoSTEALTH** [axboe/liburing#572] [Ideas]
Are there any plans to add Pause and Resume features?

**Comment by axboe** [comment]
No plans to add that, because I don't know what that would refer to. What is being paused and resumed?

**Reply by YoSTEALTH** [reply]
Hmm... lets see, in co-routine environment you can pause/suspend an task and resume it later. In terms for `io_uring` hmm... It wouldn't make much sense to pause `sqe` task since its normally one-shot `io` task, and it either does that task or not (error, cancel, ...). Maybe in terms of something like file/socket `fd` you can `open -> pause` this way it can not `read, write, close` that `fd` till you `resume`.

Lets say you are running a http server and it can only handle 100 jobs before CPU is maxed... You have received 200 connections... What you could do is accept and pause rest of the tasks till resource is freed and resume. Most cases it could be an "bad actor", now you just pause their connections only, freeing system resources and than you can deal with that problem.

Another way to think about it would to move paused task into `ttl` this way its "paused" till timer runs out and it "resumes". Note: This can already be done using io_uring `io_uring_prep_timeout*`

I noticed there was `io_uring_sqe.ioprio` I am not really sure what its for. I am just assuming you have some kind of plan to prioritize one `sqe` over other?

**Reply by YoSTEALTH** [reply]
This is very easy to do in `Python`, you just remove the co-routine object from being used in event-manager, thus giving it that suspended effect, later you inject is back to resume. I am not sure how all this would work in `C` world.

**Comment by axboe** [comment]
I see - yes I don't think this fits with the OS/C model at all, as you can't really pause access to an fd. I suspect this is a higher level primitive that needs to be applied as such, not something io_uring can directly sppport.

ioprio is passed down the stack and already works for things like storage, it's not meant to be used in io_uring core.

**Comment by YoSTEALTH** [comment]
> I see - yes I don't think this fits with the OS/C model at all, as you can't really pause access to an fd.

Ok, dealing with `fd` might have been too much of a stress.


> ioprio is passed down the stack and already works for things like storage, it's not meant to be used in io_uring core.

I see.

> I suspect this is a higher level primitive that needs to be applied as such, not something io_uring can directly support.

Its not something I would bug you to implement since its easy to do this in higher languages.

That being said I don't think its impossible either. Lets think about it... now that `fd` method is no go, how about dealing with `sqe/cqe`

------

Lets say you submit `5` entries:

`sqe->user_data` = `[1, 2, 3, 4, 5]`

You want to pause `4` but `sqe` is already been submitted.

The task is either not processed (still in sqe) or finished (already in cqe).

In `io_uring` you would remove `4` either from `sqe` or `cqe` and put it in temp holder.

Later you resume `4`, now `io_uring` take `4` from temp holder and puts it back into sqe queue or returns the already finished `cqe`.

**Comment by axboe** [comment]
My point is that since it's easy to do elsewhere, adding extra code and tracking for this in io_uring would then just a cost that everyone else would have to pay. Hence I don't think that's a worthy tradeoff. The features that belong in io_uring are things you can't easily do higher up.
